### PR TITLE
ff: add WithConfigFileVia option

### LINKED ISSUE #70: Not sure how to properly to share config file between root flag and subcommands.
I want something like this to work: `mycmd -config configfile subcommand -subarg ...` where one config file has values for both root and any subcommands. 

Having one config file per subcommand with it's own `-config` argument is too messy and user unfriendly to consider, the one root level `-config` should be enough and the config file should preferably not have to be reread multiple times.

I'm not sure how to express this in a clean way using ffcli.

### PR DESCRIPTION
**Pull request by thomasf**
as discussed in #70 

closes #70

### CODE REVIEW

**Comment by thomasf** [comment]
I see a bug. will amend

**Comment by thomasf** [comment]
WithConfigFileVia could maybe be named WithConfigFileVar because it is a similar pattern to how FlagSet.StringVar(...) is named, not entirely the same thing though.

File: parse.go [hunk 1]
```diff
@@ -118,6 +136,7 @@ func Parse(fs *flag.FlagSet, args []string, options ...Option) error {
 // Context contains private fields used during parsing.
 type Context struct {
 	configFile             string
```
Comment (peterbourgon) [inline parse.go:138]: I think you can refactor WithConfigFile to be sugar for WithConfigFileVia, which would let you remove this field and leave Parse essentially untouched.

Comment (thomasf) [reply parse.go]: done

**Comment by thomasf** [comment]
Maybe the test is a bit misplaced (ffcli instead of ff package) but it does ensure that my particular use case works.

File: parse.go [hunk 2]
```diff
@@ -67,14 +67,15 @@ func Parse(fs *flag.FlagSet, args []string, options ...Option) error {
 	})
 
 	// Third priority: config file (host).
-	if c.configFile == "" && c.configFileFlagName != "" {
+	if (c.configFileVia == nil || *c.configFileVia == "") && c.configFileFlagName != "" {
```
Comment (peterbourgon) [inline parse.go:70]:
```suggestion
	var configFile string
	if c.configFileVia != nil {
		configFile = *c.configFileVia
	}
	if configFile == "" && c.configFileFlagName != "" {
```

File: parse.go [hunk 3]
```diff
@@ -67,14 +67,15 @@ func Parse(fs *flag.FlagSet, args []string, options ...Option) error {
...
-			c.configFile = f.Value.String()
+			v := f.Value.String()
+			c.configFileVia = &v
 		}
 	}
 
-	if parseConfig := c.configFile != "" && c.configFileParser != nil; parseConfig {
-		f, err := os.Open(c.configFile)
+	if parseConfig := c.configFileVia != nil && *c.configFileVia != "" && c.configFileParser != nil; parseConfig {
+		f, err := os.Open(*c.configFileVia)
```
Comment (peterbourgon) [inline parse.go:78]:
```suggestion
	if parseConfig := configFile != "" && c.configFileParser != nil; parseConfig {
		f, err := os.Open(configFile)
```

File: parse_test.go [hunk 4]
```diff
@@ -254,3 +257,58 @@ func TestParseConfigFile(t *testing.T) {
...
+	t.Parallel()
+
+	rootFS := flag.NewFlagSet("root", flag.ContinueOnError)
+	config := rootFS.String("config-file", "", "")
+	iFlag := rootFS.Int("i", 0, "")
+	sFlag := rootFS.String("s", "", "")
+
+	subFS := flag.NewFlagSet("subcommand", flag.ContinueOnError)
+	dFlag := subFS.Duration("d", time.Second, "")
+	bFlag := subFS.Bool("b", false, "")
```
Comment (peterbourgon) [inline parse_test.go:271]:
```suggestion
	var (
		rootFS = flag.NewFlagSet("root", flag.ContinueOnError)
		config = rootFS.String("config-file", "", "")
		i     = rootFS.Int("i", 0, "")
		s     = rootFS.String("s", "", "")

		subFS = flag.NewFlagSet("subcommand", flag.ContinueOnError)
		d     = subFS.Duration("d", time.Second, "")
		b     = subFS.Bool("b", false, "")
	)
```

File: parse_test.go [hunk 5]
```diff
@@ -254,3 +257,58 @@ func TestParseConfigFile(t *testing.T) {
...
+		t.Error(*dFlag)
+	}
+	if *sFlag != "bar" {
+		t.Error(*sFlag)
+	}
+	if *iFlag != 99 {
+		t.Error(*iFlag)
+	}
+
+}
```
Comment (peterbourgon) [inline parse_test.go:314]:
```suggestion
	if want, have := time.Hour, *dFlag; want != have {
		t.Errorf("d: want %v, have %v", want, have)
	}
	if want, have := false, *bFlag; want != have {
		t.Errorf("b: want %v, have %v", want, have)
	}
	if want, have := "bar", *sFlag; want != have {
		t.Errorf("s: want %q, have %q", want, have)
	}
	if want, have := 99, *iFlag; want != have {
		t.Errorf("i: want %d, have %d", want, have)
	}
}
```

Comment (thomasf) [reply parse_test.go]: don't really see the need for this, IMO it just adds noise but it's your package so I've made the change anyway

Comment (peterbourgon) [reply parse_test.go]: Well, it did allow me to catch and fix the bug below :)

```go
	if !*bFlag {
		t.Error(*dFlag)
	}
```

File: parse_test.go [hunk 6]
```diff
@@ -1,11 +1,14 @@
...
 
 import (
+	"context"
+	"flag"
 	"os"
 	"testing"
 	"time"
 
 	"github.com/peterbourgon/ff/v3"
+	"github.com/peterbourgon/ff/v3/ffcli"
```
Comment (peterbourgon) [inline parse_test.go:11]: :\ I'll refactor this later.

**Review by peterbourgon — APPROVED** [review]
Cool!

**Pull request merged**